llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -2960,6 +2967,197 @@ static bool validateAndCostRequiredSelects(BasicBlock *BB, BasicBlock *ThenBB,
+///   %cond.not = xor i1 %cond, true
+///   %vcond.not = bitcast i1 %cond.not to <1 x i>
+///   call void @llvm.masked.store.v1i32.p0
+///              (<1 x i32> <i32 1>, ptr %q, i32 4, <1x i1> %vcond.not)
+/// \endcode
+///
+/// \returns true if any load/store is hosited.
+///
+/// Note that this tranform should be run
+/// * before SpeculativelyExecuteBB so that the latter can have more chance.
+/// * after hoistCommonCodeFromSuccessors to ensure unconditional loads/stores
+///   are handled first.
+bool SimplifyCFGOpt::hoistLoadStoreWithCondFaultingFromSuccessors(
+    BasicBlock *BB) {
+  if (!HoistLoadsStoresWithCondFaulting)
+    return false;
+
+  auto *BI = dyn_cast<BranchInst>(BB->getTerminator());
+  if (!BI || !BI->isConditional())
+    return false;
+
+  BasicBlock *IfTrueBB = BI->getSuccessor(0);
+  BasicBlock *IfFalseBB = BI->getSuccessor(1);
+
+  // If either of the blocks has it's address taken, then we can't do this fold,
+  // because the code we'd hoist would no longer run when we jump into the block
+  // by it's address.
+  for (auto *Succ : {IfTrueBB, IfFalseBB})
+    if (Succ->hasAddressTaken())
+      return false;
+
+  // Collect hoisted loads/stores.
+  SmallSetVector<Instruction *, 4> HoistedInsts;
+  // Not hoist load/store if
+  // 1. target does not have corresponding conditional faulting load/store.
+  // 2. it's volatile or atomic.
+  // 3. there is a load/store that can not be hoisted in the same bb.
+  // 4. there is a non-load/store that may have side effects in the same bb.
+  // 5. any operand of it does not dominate the branch.
+  // 6. it's a store and a memory read is skipped.
+  auto HoistInstsInBB = [&](BasicBlock *BB) {
+    bool SkipMemoryRead = false;
+    // A more efficient way to check domination. An operand dominates the
+    // BranchInst if
+    // 1. it's not defined in the same bb as the instruction.
+    // 2. it's to be hoisted.
+    //
+    // b/c BB is only predecessor and BranchInst does not define any value.
+    auto OpsDominatesBranch = [&](Instruction &I) {
+      return llvm::none_of(I.operands(), [&](Value *Op) {
+        if (auto *J = dyn_cast<Instruction>(Op)) {
+          if (HoistedInsts.contains(J))
+            return false;
+          if (J->getParent() == I.getParent())
+            return true;
+        }
+        return false;
+      });
+    };
+    for (auto &I : *BB) {
+      auto *LI = dyn_cast<LoadInst>(&I);
+      auto *SI = dyn_cast<StoreInst>(&I);
+      if (LI || SI) {
+        // a load from alloca is always safe.
+        if (isLoadFromAlloca(I)) {
+          HoistedInsts.insert(&I);
+          continue;
+        }
+        auto *Type = LI ? I.getType() : I.getOperand(0)->getType();
+        bool IsSimple = (LI && LI->isSimple()) || (SI && SI->isSimple());
+        if (!TTI.hasConditionalFaultingLoadStoreForType(Type) || !IsSimple ||
+            !OpsDominatesBranch(I))
+          return false;
+        if (SI && SkipMemoryRead)
```

Comment:
IIUC, this is essentially a conservative aliasing check?

Reply:
If so, does a store not essentially do the same thing for loads?

Reply:
Yes. Will add it to comments.

Reply:
> Will add it to comments.

Done.

> does a store not essentially do the same thing for loads?

 A store is different from load to call intrinsic with attributes like "readonly nounwind willreturn". 
 See test load_skip_memory_read/not_store_skip_memory_read in hoist-load-store-with-cf.ll

Reply:
AFAICT those tests for not hoisting a store when there are loads, but AFAICT, you will still hoist loads if there are stores. Take the following:

BB0:
  p = alloc
  do_stuff0(p)
  br BB1/BB2

BB1:
   store p
   do_stuff1(p)
   load p
   do_stuff2(p)
BB2:
   load p
   do_stuff3(p)


As I understand it, the 'load p' gets hoisted.

Reply:
Your understanding is right. If  `do_stuff1`, `do_stuff2` and `do_stuff3` do not have side effects,  `load p` and `store p` in both BB1/BB2 will be hoisted.  Is there any issue?

Reply:
I got your idea. Added test load_after_store for this case.

Reply:
Oh I see, so the argument is that since the store will also be hoisted its okay.
Can you also add a test like:

BB0:
p2 = some_ptr
p = alloc
do_stuff0(p)
br BB1/BB2

BB1:
load p2
store p
do_stuff1(p)
load p
do_stuff2(p)
BB2:
load p
do_stuff3(p)

I think you might hit a case where the store doesn't hoist (b.c maybe aliases w/ `load p2`) but the `load p` does.


As an aside, there should probably be some hoist limit on this. If you are hoisting more than several loads/stores that are only on one side of the condition, a branch is probably cheaper...

Reply:
> I think you might hit a case where the store doesn't hoist

See comment at line 3057 in SimplifyCFG.cpp.  We do not hoist if there is a load/store that can not be hoisted in the same bb. Test not_hoistable_store checks the logic.

> a branch is probably cheaper

Plan to do it but not in this patch. This patch is big enough.

Reply:
> > I think you might hit a case where the store doesn't hoist
> 
> See comment at line 3057 in SimplifyCFG.cpp. We do not hoist if there is a load/store that can not be hoisted in the same bb. Test not_hoistable_store checks the logic.

I see, although do we really want this to be all-or-nothing?

> 
> > a branch is probably cheaper
> 
> Plan to do it but not in this patch. This patch is big enough.

Agreed. Was thinking in fact, could we split the SimplfyCFG stuff into a seperate one?
Or at the very least this patch should have [SimplifyCFG] tag.

Reply:
At this point, yes. All or nothing is a good start. Just added the [SimplifyCFG] tag.

Reply:
If possible, I would like this to be a separate PR.